File: app/src/main/java/com/vwoom/timelapsegallery/adapters/DetailsAdapter.java
package com.vwoom.timelapsegallery.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.vwoom.timelapsegallery.R;
import com.vwoom.timelapsegallery.database.entry.PhotoEntry;
import com.vwoom.timelapsegallery.database.entry.ProjectEntry;
import com.vwoom.timelapsegallery.database.view.Project;
import com.vwoom.timelapsegallery.utils.FileUtils;

import java.io.File;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class DetailsAdapter extends RecyclerView.Adapter<DetailsAdapter.DetailsAdapterViewHolder> {

    private final static String TAG = DetailsAdapter.class.getSimpleName();

    private List<PhotoEntry> mPhotos;
    private Project mProject;
    private final DetailsAdapterOnClickHandler mClickHandler;
    private PhotoEntry mCurrentPhoto;

    public interface DetailsAdapterOnClickHandler {
        void onClick(PhotoEntry clickedPhoto);
    }

    public DetailsAdapter(DetailsAdapterOnClickHandler handler){ mClickHandler = handler;}

    public class DetailsAdapterViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        @BindView(R.id.detail_thumbnail) ImageView mDetailThumbnail;
        @BindView(R.id.selection_indicator) View mSelectionIndicator;

        public DetailsAdapterViewHolder(View view){
            super(view);
            ButterKnife.bind(this, view);
            view.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            int adapterPosition = getAdapterPosition();
            PhotoEntry clickedPhoto = mPhotos.get(adapterPosition);
            mCurrentPhoto = clickedPhoto;
            mClickHandler.onClick(clickedPhoto);
        }
    }

    @NonNull
    @Override
    public DetailsAdapterViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        int layoutIdForGridItem = R.layout.detail_recyclerview_item;
        LayoutInflater inflater = LayoutInflater.from(context);
        boolean shouldAttachToParentImmediately = false;
        View view = inflater.inflate(layoutIdForGridItem, parent, shouldAttachToParentImmediately);
        return new DetailsAdapterViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull DetailsAdapterViewHolder holder, int position) {
        Context context = holder.itemView.getContext();
        PhotoEntry currentPhoto = mPhotos.get(position);
        String photo_path = FileUtils.getPhotoUrl(context, mProject, currentPhoto);
        File f = new File(photo_path);

        // TODO (update) dynamically resize detail view
        Glide.with(context)
                .load(f)
                .centerCrop()
                .into(holder.mDetailThumbnail);

        if (mPhotos.get(position) == mCurrentPhoto){
            holder.mSelectionIndicator.setVisibility(View.VISIBLE);
        } else {
            holder.mSelectionIndicator.setVisibility(View.INVISIBLE);
        }
    }

    @Override
    public int getItemCount() {
        if (mPhotos == null) return 0;
        return mPhotos.size();
    }

    public void setPhotoData(List<PhotoEntry> photoData, Project project){
        mPhotos = photoData;
        mProject = project;
        notifyDataSetChanged();
    }


    public void setCurrentPhoto(PhotoEntry photo){
        mCurrentPhoto = photo;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/vwoom/timelapsegallery/adapters/DetailsAdapter.kt
package com.vwoom.timelapsegallery.adapters

import android.content.Context
import android.os.Environment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import com.bumptech.glide.Glide
import com.vwoom.timelapsegallery.R
import com.vwoom.timelapsegallery.adapters.DetailsAdapter
import com.vwoom.timelapsegallery.adapters.DetailsAdapter.DetailsAdapterViewHolder
import com.vwoom.timelapsegallery.data.entry.PhotoEntry
import com.vwoom.timelapsegallery.data.view.Project
import com.vwoom.timelapsegallery.utils.FileUtils
import java.io.File

class DetailsAdapter(private val mClickHandler: DetailsAdapterOnClickHandler, context: Context) : RecyclerView.Adapter<DetailsAdapterViewHolder>() {
    private var mPhotos: List<PhotoEntry>? = null
    private var mProject: Project? = null
    private var mCurrentPhoto: PhotoEntry? = null
    private val mExternalFilesDir: File?

    interface DetailsAdapterOnClickHandler {
        fun onClick(clickedPhoto: PhotoEntry?)
    }

    inner class DetailsAdapterViewHolder(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener {
        @BindView(R.id.detail_thumbnail)
        var mDetailThumbnail: ImageView? = null
        @BindView(R.id.selection_indicator)
        var mSelectionIndicator: View? = null

        override fun onClick(view: View) {
            val adapterPosition = adapterPosition
            val clickedPhoto = mPhotos!![adapterPosition]
            mCurrentPhoto = clickedPhoto
            mClickHandler.onClick(clickedPhoto)
        }

        init {
            ButterKnife.bind(this, view)
            view.setOnClickListener(this)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DetailsAdapterViewHolder {
        val context = parent.context
        val layoutIdForGridItem = R.layout.detail_recyclerview_item
        val inflater = LayoutInflater.from(context)
        val shouldAttachToParentImmediately = false
        val view = inflater.inflate(layoutIdForGridItem, parent, shouldAttachToParentImmediately)
        return DetailsAdapterViewHolder(view)
    }

    override fun onBindViewHolder(holder: DetailsAdapterViewHolder, position: Int) {
        val context = holder.itemView.context
        val currentPhoto = mPhotos!![position]
        val photo_path = FileUtils.getPhotoUrl(mExternalFilesDir, mProject, currentPhoto)
        val f = File(photo_path)
        // TODO (update) dynamically resize detail view
        Glide.with(context)
                .load(f)
                .centerCrop()
                .into(holder.mDetailThumbnail!!)
        if (mPhotos!![position] === mCurrentPhoto) {
            holder.mSelectionIndicator!!.visibility = View.VISIBLE
        } else {
            holder.mSelectionIndicator!!.visibility = View.INVISIBLE
        }
    }

    override fun getItemCount(): Int {
        return if (mPhotos == null) 0 else mPhotos!!.size
    }

    fun setPhotoData(photoData: List<PhotoEntry>?, project: Project?) {
        mPhotos = photoData
        mProject = project
        notifyDataSetChanged()
    }

    fun setCurrentPhoto(photo: PhotoEntry?) {
        mCurrentPhoto = photo
        notifyDataSetChanged()
    }

    companion object {
        private val TAG = DetailsAdapter::class.java.simpleName
    }

    init {
        mExternalFilesDir = context.getExternalFilesDir(Environment.DIRECTORY_PICTURES)
    }
}